server.go

```diff
@@ -24,113 +26,140 @@ type Route struct {
+		cmd := exec.Command("roc", "build", scriptPath)
 		// --optimize runs slow on Roc PG for now
 		// cmd := exec.Command("roc", "build", "--optimize", script_path)
-		err := cmd.Run()
-		if err != nil {
-			log.Fatalf("ERROR: Unable to build %s: %s", script_path, err.Error())
+		if err := cmd.Run(); err != nil {
+			return fmt.Errorf("unable to build %s: %w", scriptPath, err)
 		}
 
 		// Check that the executable exists with the expected name
-		if _, err := os.Stat(binary_path); os.IsNotExist(err) {
-			log.Fatalf("ERROR: Expected binary %s does not exist", binary_path)
+		if _, err := os.Stat(binaryPath); errors.Is(err, fs.ErrNotExist) {
+			return fmt.Errorf("expected binary %s does not exist", binaryPath)
 		}
-
 	}
 
 	http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
-
 		start := time.Now()
 
-		handleHTTPRequest(w, r, routes, cgi_scripts_dir)
+		if err := handleHTTPRequest(w, r, routes, cgiScriptsDir); err != nil {
+			statusCode := 500
+			var errStatusCode statusCodeError
+			if errors.As(err, &errStatusCode) {
+				statusCode = int(errStatusCode)
+			}
+
+			log.Printf("Error: %v", err)
+			w.WriteHeader(statusCode)
+			return
+		}
 
 		elapsed := time.Since(start)
 
 		log.Printf("%s %s %d ms\n", r.Method, r.RequestURI, elapsed.Milliseconds())
 	})
 
 	log.Printf("INFO: Listening on port 8080\n")
-	http.ListenAndServe(":8080", nil)
+	if err := http.ListenAndServe(":8080", nil); err != http.ErrServerClosed {
+		return fmt.Errorf("HTTP Server failed: %w", err)
+	}
+
+	return nil
 }
 
-func readAndUnmarshalRoutes(routeFile string) Routes {
+func readAndUnmarshalRoutes(routeFile string) (Routes, error) {
 	data, err := os.ReadFile(routeFile)
 	if err != nil {
-		log.Fatalf("ERROR: Unable to read file: %s", err.Error())
+		return Routes{}, fmt.Errorf("unable to read file: %w", err)
 	}
 
 	var routes Routes
 	err = yaml.Unmarshal(data, &routes)
 	if err != nil {
-		log.Fatalf("ERROR: Unable to decode YAML: %s", err.Error())
+		return Routes{}, fmt.Errorf("unable to decode YAML: %w", err)
 	}
 
-	sort.Sort(ByPathLength(routes.Routes))
+	sort.Sort(byPathLength(routes.Routes))
 
-	return routes
+	return routes, nil
 }
 
-func findRoute(r *http.Request, routes Routes) *Route {
+func findRoute(r *http.Request, routes Routes) (*Route, bool) {
 	for _, route := range routes.Routes {
 		params := parseURLParameters(r.URL.Path, route.Path)
 		if params != nil && r.Method == route.Method {
 			route.Params = params
-			return &route
+			return &route, true
 		}
 	}
 
-	return nil
+	return nil, false
 }
 
-func handleHTTPRequest(w http.ResponseWriter, r *http.Request, routes Routes, cgi_scripts_dir string) {
-	route := findRoute(r, routes)
-	if route == nil {
-		http.Error(w, "no script found for route", http.StatusBadRequest)
-		return
+func handleHTTPRequest(w http.ResponseWriter, r *http.Request, routes Routes, cgiScriptsDir string) error {
+	route, ok := findRoute(r, routes)
+	if !ok {
+		return errors.Join(
```

Comment:
errors.Join was added in go 1.20 (current release) If you want to support older versions of go, another solution has to be used.